js/impress.js

```diff
@@ -659,6 +705,64 @@
         return lib;
     };
 
+    // `addPreInitPlugin` allows plugins to register a function that should
+    // be run (synchronously) at the beginning of init, before
+    // impress().init() itself executes.
+    impress.addPreInitPlugin = function( plugin, weight ) {
+        weight = parseInt( weight ) || 10;
+        if ( weight <= 0 ) {
+            throw "addPreInitPlugin: weight must be a positive integer";
+        }
+
+        if ( preInitPlugins[ weight ] === undefined ) {
+            preInitPlugins[ weight ] = [];
+        }
+        preInitPlugins[ weight ].push( plugin );
+    };
+
+    // Called at beginning of init, to execute all pre-init plugins.
+    var execPreInitPlugins = function( root ) { //jshint ignore:line
+        for ( var i = 0; i < preInitPlugins.length; i++ ) {
+            var thisLevel = preInitPlugins[ i ];
+            if ( thisLevel !== undefined ) {
+                for ( var j = 0; j < thisLevel.length; j++ ) {
+                    thisLevel[ j ]( root );
+                }
+            }
+        }
+    };
+
+    // `addPreStepLeavePlugin` allows plugins to register a function that should
+    // be run (synchronously) at the beginning of goto()
+    impress.addPreStepLeavePlugin = function( plugin, weight ) { //jshint ignore:line
```

Comment:
I don't like the fact that we are adding a lot of (kind of) internal api to global `impress` namespace, as it seems polluting it a bit...
I know that these methods need to be globally accessible, but maybe wrapping them in some `plugin` namespace would work? Like `impress.plugin.addPreStepLeavePlugin` etc?

As previously - not something I want to be changed now (knowing the dependency on existing plugins), but something to consider when we make new plugin API released and public.

Reply:
Actually, I could make such change. The nr of plugins that use this api is small, in fact I think they are contained in this PR, so it is feasible to change.

I should note, in defense of this choice, that hey aren't added to the api returned by impress(). (Which would actually be incorrect anyway, since these are called before impress().init()). So they are kind of hidden aside already, and there are only two.


So you're suggesting:

impress.plugin.addPreInitPlugin()
impress.plugin.addPreStepLeavePlugin()

I actually think that's a bit confusing. These methods are offered by core impress.js. When I look at the above, it looks more like they *are* plugins, which is not the case.

But let me know if you insist, this is a change that can be done in this PR.

Reply:
I know that they are not part of api returned by `impress()` and it's good.
And also, now looking at this I agree that it may be interpreted as kind of a plugin, so I'm fine with leaving it as it is. Maybe later we can have a look on what all methods are on `impress` object and see if it makes sense to move or rename them somehow.

Reply:
I go for @henrikingo on this one. The methods are built to create plugins so there's no reason to be anywhere else other than the core.